### PR TITLE
feat: convert xml to protobuf IDL and json targets

### PR DESCRIPTION
**Pull request by huieric**

### CODE REVIEW

File: internal/confgen/confgen.go [hunk 1]
```diff
@@ -23,13 +23,29 @@ type Generator struct {
...
+	Input *options.InputOption // Input settings.
+	Header *options.HeaderOption // header settings.
 }
 
 var specialMessageMap = map[string]int{
 	"google.protobuf.Timestamp": 1,
 	"google.protobuf.Duration":  1,
 }
 
+func NewGenerator(protoPackage, indir, outdir string, setters ...options.Option) (*options.Options, *Generator) {
```
Comment (wenchy) [inline internal/confgen/confgen.go]: It's not good to return options, as the options is input parameters.

File: internal/importer/xml.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	"github.com/Wenchy/tableau/internal/xlsxgen"
+	"github.com/Wenchy/tableau/options"
+	"github.com/Wenchy/tableau/proto/tableaupb"
+	"github.com/antchfx/xmlquery"
+	"github.com/antchfx/xpath"
+	"github.com/pkg/errors"
+	"github.com/iancoleman/strcase"
+)
+
+type XmlImporter struct {
```
Comment (wenchy) [inline internal/importer/xml.go]: naming: `XmlImporter` -> `XMLImporter`

File: internal/importer/xml.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+	attrValRegexp := regexp.MustCompile(`"\S+"`)
+	nudeRegexp := regexp.MustCompile(`<([A-Za-z0-9]+)>`)
+	keywordRegexp := regexp.MustCompile(`([</]+)@([A-Z]+)`)
+	replaced_str := attrValRegexp.ReplaceAllStringFunc(string(buf), func(s string) string {
+		var buf bytes.Buffer
+		xml.EscapeText(&buf, []byte(s[1:len(s)-1]))
+		return fmt.Sprintf("\"%s\"", buf.String())
+	})
+	replaced_str = nudeRegexp.ReplaceAllString(replaced_str, `<$1 TableauPlaceholder="0">`)
+	replaced_str = keywordRegexp.ReplaceAllString(replaced_str, `$1$2`)
```
Comment (wenchy) [inline internal/importer/xml.go]: naming convention: `replaced_str` -> `replacedStr`

File: internal/importer/xml.go [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+	}
+	sheet := NewSheet(root.LocalName(), rows)
+	sheet.Meta = &tableaupb.SheetMeta{
+		Sheet: root.LocalName(),
+		Alias: root.LocalName(),
+		Nameline: 1,
+		Typeline: 1,
+		Nested: true,
+	}
+	x.sheetMap[root.LocalName()] = sheet
```
Comment (wenchy) [inline internal/importer/xml.go:123]: treated as only one sheet in a XML file?

File: internal/importer/xml.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+			for tmpCusor := cursor; tmpCusor < len(metaSheet.Rows); tmpCusor++ {
+				metaSheet.Cell(tmpCusor, colName).Data = attrValue
+			}
+		}
+
+		curType := metaSheet.GetColType(colName)
+		matches := types.MatchStruct(curType)
+		// 1. <TABLEAU>
+		// 2. type not set
+		// 3. {Type}int32 -> [Type]int32
```
Comment (wenchy) [inline internal/importer/xml.go]: Need this?

File: internal/importer/xml.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+		needChangeType := isMeta || curType == "" || (len(matches) > 0 && repeated)
+		// 1. new struct(list), not subsequent
+		// 2. {Type}int32 -> [Type]int32
+		setKeyedType := !strings.HasPrefix(lastColName, prefix) || (len(matches) > 0 && repeated)
+		if needChangeType {
+			if matches := types.MatchMap(t); len(matches) == 3 {
+				if !types.IsScalarType(matches[1]) && len(types.MatchEnum(matches[1])) == 0 {
+					return fmt.Errorf("%s is not scalar type in node %s attr %s type %s", matches[1], nav.LocalName(), attrName, t)
```
Comment (wenchy) [inline internal/importer/xml.go]: errors: replace all `fmt.Errorf` to `errors.Errorf`
refetr: https://github.com/pkg/errors

File: internal/protogen/exporter.go [hunk 7]
```diff
@@ -172,7 +173,7 @@ func (x *sheetExporter) export() error {
 	return nil
 }
 
-func (x *sheetExporter) exportField(depth int, tagid int, field *tableaupb.Field) error {
+func (x *sheetExporter) exportField(depth int, tagid int, field *tableaupb.Field, prefix string) error {
```
Comment (wenchy) [inline internal/protogen/exporter.go:173]: There is no need to care about **prefix**, as it was already nested.

File: internal/protogen/parser.go [hunk 8]
```diff
@@ -85,6 +84,9 @@ func (p *bookParser) parseSubField(field *tableaupb.Field, header *sheetHeader,
 	cursor, ok := p.parseField(subField, header, cursor, prefix, nested)
 	if ok {
 		field.Fields = append(field.Fields, subField)
+		if field.Options.Layout == tableaupb.Layout_LAYOUT_HORIZONTAL {
+			field.Options.ListMaxLen /= int32(len(field.Fields))
+		}
```
Comment (wenchy) [inline internal/protogen/parser.go]: ?

File: internal/protogen/parser.go [hunk 9]
```diff
@@ -412,8 +422,8 @@ func (p *bookParser) parseStructField(field *tableaupb.Field, header *sheetHeade
 		field.Type, field.TypeDefined = p.parseType(elemType)
 		field.Name = strcase.ToSnake(field.Type)
 		// index := len(field.Type)
-		// structName := trimmedNameCell[:index]
-		structName := field.Type
+		index := strings.Index(trimmedNameCell, field.Type) + len(field.Type)
```
Comment (wenchy) [inline internal/protogen/parser.go]: intent?

File: internal/protogen/protogen.go [hunk 10]
```diff
@@ -34,9 +34,80 @@ type Generator struct {
...
+
+		FilenameWithSubdirPrefix: opts.Output.FilenameWithSubdirPrefix,
+		FilenameSuffix:           opts.Output.FilenameSuffix,
+
+		Imports: opts.Imports,
+		Header:         opts.Header,
+		InputOpts: opts.Input,
+	}
+	return opts, g
+}
```
Comment (wenchy) [inline internal/protogen/protogen.go]: 1. format!
2. intent?

File: internal/protogen/protogen.go [hunk 11]
```diff
@@ -46,11 +117,6 @@ func (gen *Generator) Generate() error {
 	return gen.generate(gen.InputDir)
 }
 
-func (gen *Generator) GenOneWorkbook(relativeWorkbookPath string) error {
-	absPath := filepath.Join(gen.InputDir, relativeWorkbookPath)
-	return gen.convertWorkbook(filepath.Dir(absPath), filepath.Base(absPath))
-}
-
```
Comment (wenchy) [inline internal/protogen/protogen.go]: ?

File: internal/protogen/protogen.go [hunk 12]
```diff
@@ -34,9 +34,80 @@ type Generator struct {
...
+	case options.Excel:
+		return ".xlsx"
+	case options.CSV:
+		return ".csv"
+	case options.XML:
+		return ".xml"
+	}
+	// default excel
+	return ".xlsx"
+}
```
Comment (wenchy) [inline internal/protogen/protogen.go]: boring stuff! You'd better use Format type to check the suffix.

File: internal/protogen/protogen.go [hunk 13]
```diff
@@ -64,19 +130,18 @@ func (gen *Generator) generate(dir string) error {
...
-			continue
 		}
 
 		if strings.HasPrefix(entry.Name(), "~$") {
 			// ignore xlsx temp file named with prefix "~$"
 			continue
 		}
 		// atom.Log.Debugf("generating %s, %s", entry.Name(), filepath.Ext(entry.Name()))
-		if filepath.Ext(entry.Name()) != ".xlsx" {
+		if filepath.Ext(entry.Name()) != gen.GetSuffix() {
```
Comment (wenchy) [inline internal/protogen/protogen.go]: use Format type to check

File: options/options.go [hunk 14]
```diff
@@ -31,6 +35,10 @@ type HeaderOption struct {
 	Typeline int32
 }
 
+type CustomRowData struct {
+	Default string `json:"default"`
+}
+
```
Comment (wenchy) [inline options/options.go]: What for?

File: proto/tableau/protobuf/tableau.proto [hunk 15]
```diff
@@ -65,6 +65,9 @@ message FieldOptions {
   int32 min = 11;     // min value
   int32 max = 12;     // max value
   string range = 13;  // format like set description: [1,10], (1,10], [1,10), [1,~]
+
+  string default = 14; // field default value
+  int32 list_max_len = 15; // non-incell list max len
```
Comment (wenchy) [inline proto/tableau/protobuf/tableau.proto]: not used?

File: tableau.go [hunk 16]
```diff
@@ -33,22 +26,9 @@ func Excel2Conf(protoPackage, indir, outdir string, setters ...options.Option) {
...
-		FilenameWithSubdirPrefix: opts.Output.FilenameWithSubdirPrefix,
-		FilenameSuffix:           opts.Output.FilenameSuffix,
-
-		Header:  opts.Header,
-		Imports: opts.Imports,
-	}
+	opts, g := protogen.NewGenerator(protoPackage, goPackage, indir, outdir, setters...)
 	atom.InitZap(opts.LogLevel)
-	atom.Log.Debugf("options inited: %+v, header: %+v, output: %+v", opts, opts.Header, opts.Output)
+	atom.Log.Infof("options inited: %+v", opts)
```
Comment (wenchy) [inline tableau.go]: log level: `Infof` -> `Debugf`

File: tableau.go [hunk 17]
```diff
@@ -73,3 +53,26 @@ func ParseMeta(indir, relWorkbookPath string) importer.Importer {
 		importer.Parser(parser),
 	)
 }
+
+// Xml2Conf converts xml files to different formatted configuration files.
+// Supported formats: json, prototext, and protowire.
+func Xml2Conf(protoPackage, indir, outdir string, setters ...options.Option) {
+	opts, g := confgen.NewGenerator(protoPackage, indir, outdir, setters...)
+	atom.InitZap(opts.LogLevel)
+	atom.Log.Infof("options inited: %+v", opts)
```
Comment (wenchy) [inline tableau.go]: log level: `Infof` -> `Debugf`

File: test/protoconf/excel/test_conf.proto [hunk 18]
```diff
@@ -130,9 +130,9 @@ message MatchConf {
...
   option (tableau.worksheet) = {name:"Loader" namerow:1 typerow:2 noterow:3 datarow:5 nameline:1 typeline:1 nested:true};
 
-  map<int32, Server> server_map = 1 [(tableau.field) = {name:"Server" key:"Type" layout:LAYOUT_VERTICAL}];
+  map<string, Server> server_map = 1 [(tableau.field) = {name:"Server" key:"Name" layout:LAYOUT_VERTICAL}];
   message Server {
-    ServerType type = 1 [(tableau.field) = {name:"Type"}];
+    string name = 1 [(tableau.field) = {name:"Name"}];
     repeated Conf conf_list = 2 [(tableau.field) = {name:"Conf" key:"Type" layout:LAYOUT_VERTICAL}];
     message Conf {
       ConfType type = 1 [(tableau.field) = {name:"Type"}];
```
Comment (wenchy) [inline test/protoconf/excel/test_conf.proto]: ?

File: test/scripts/gen_pb.sh [hunk 19]
```diff
@@ -15,14 +15,19 @@ test_outdir="./test/testpb"
...
+            --go_out="$test_outdir" \
+            --go_opt=paths=source_relative \
+            --proto_path="$1" \
+            --proto_path="$tableau_proto" \
+            "$item"
+        fi
+    done
+}
+gen_pb $test_indir/excel
+gen_pb $test_indir/xml
```
Comment (wenchy) [inline test/scripts/gen_pb.sh]: two packages: `testexcel` and `testxml`

File: test/tableau_test.go [hunk 20]
```diff
@@ -104,3 +104,37 @@ func Test_CSV2Excel(t *testing.T) {
...
+		),
+		options.Input(
+			&options.InputOption{
+				Format: options.XML,
+			},
+		),
+	)
+}
+
+func Test_Xml2JSON(t *testing.T) {
```
Comment (wenchy) [inline test/tableau_test.go]: naming: `Xml` -> `XML`

**Review by wenchy — CHANGES_REQUESTED** [review]

File: internal/protogen/exporter.go [hunk 21]
```diff
@@ -197,10 +198,11 @@ func (x *sheetExporter) exportField(depth int, tagid int, field *tableaupb.Field
 
 	if !field.TypeDefined && field.Fields != nil {
 		// iff field is a map or list and message type is not imported.
-		nestedMsgName := field.Type
+		msgName := field.Type
 		if field.MapEntry != nil {
-			nestedMsgName = field.MapEntry.ValueType
+			msgName = field.MapEntry.ValueType
 		}
+		nestedMsgName := prefix + "/" + msgName
```
Comment (wenchy) [inline internal/protogen/exporter.go]: `"/"` -> `"."`

File: internal/protogen/exporter.go (hunk 7 above)
Comment (huieric) [reply internal/protogen/exporter.go:173]: The same messages in different trees should be exported to different message.
So we need the `prefix` to provide their parents name and construct a nest message name.

File: test/scripts/gen_pb.sh (hunk 19 above)
Comment (huieric) [reply test/scripts/gen_pb.sh]: if excel and xml are two packages, there exists a problem. Some common messages and enum predefined in `common.proto`, `time.proto`, etc. cannot stay in one package, which means some enums may need to define twice.

File: internal/protogen/protogen.go (hunk 13 above)
Comment (huieric) [reply internal/protogen/protogen.go]: What is `Format type`?

File: internal/protogen/protogen.go (hunk 12 above)
Comment (huieric) [reply internal/protogen/protogen.go]: In this way how can i skip excel file when i only want to generate xml proto?

File: internal/importer/xml.go (hunk 4 above)
Comment (huieric) [reply internal/importer/xml.go:123]: For simplicity we just view one xml file as one sheet in this version.

**Pull request merged**